client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java

```diff
@@ -0,0 +1,133 @@
+            if (cmdSenderResult.succeeded()) {
+                final DelegatedCommandSender sender = cmdSenderResult.result();
+                sender.sendCommandMessage(command, commandContext.getTracingContext()).setHandler(sendResult -> {
+                    if (sendResult.succeeded()) {
+                        // send succeeded - handle outcome
+                        final ProtonDelivery delegatedMsgDelivery = sendResult.result();
+                        LOG.trace("command for device {} sent to downstream peer; remote state of delivery: {}",
+                                deviceId, delegatedMsgDelivery.getRemoteState());
+                        applyDelegatedMessageDeliveryResultToCommandContext(delegatedMsgDelivery, commandContext);
+                    } else {
+                        // failed to send message
+                        LOG.error("failed to send command message to downstream peer", sendResult.cause());
+                        TracingHelper.logError(commandContext.getCurrentSpan(),
+                                "failed to send command message to downstream peer: " + sendResult.cause());
+                        commandContext.release();
+                    }
+                });
+            } else {
+                // failed to create sender
+                LOG.error("failed to create sender for sending command message to downstream peer", cmdSenderResult.cause());
+                TracingHelper.logError(commandContext.getCurrentSpan(),
+                        "failed to create sender for sending command message to downstream peer: " + cmdSenderResult.cause());
+                commandContext.release();
+            }
+        });
+    }
+
+    private void applyDelegatedMessageDeliveryResultToCommandContext(final ProtonDelivery delegatedMsgDelivery, final CommandContext commandContext) {
+        switch (delegatedMsgDelivery.getRemoteState().getType()) {
+            case Accepted:
+                commandContext.accept();
+                break;
+            case Rejected:
+                final Rejected rejected = (Rejected) delegatedMsgDelivery.getRemoteState();
+                commandContext.reject(rejected.getError());
+                break;
+            case Modified:
+                final Modified modified = (Modified) delegatedMsgDelivery.getRemoteState();
+                commandContext.modify(modified.getDeliveryFailed(), modified.getUndeliverableHere(), 0);
+                break;
+            case Released:
+                commandContext.release();
+                break;
+            default:
+                LOG.warn("got unexpected delivery outcome; remote state: {}", delegatedMsgDelivery.getRemoteState());
+                TracingHelper.logError(commandContext.getCurrentSpan(), "got unexpected delivery outcome; remote state: "
+                        + delegatedMsgDelivery.getRemoteState());
+                commandContext.getCurrentSpan().finish();
```

Comment:
My understanding is that at this point the application that sent the message is still expecting a disposition frame indicating the outcome of the transfer, isn't it? IMHO we need to invoke any of the `CommandContext`'s methods in order to finish the original message transfer here, i.e. even if we received an unexpected outcome, we should send a disposition with a supported outcome to the application, don't you think?.

It also feels strange that we finish the Span in this case but not in the other cases ...

Reply:
Right, I've put a `commandContext.release()` in there. 
Besides the transaction related outcomes (`Declared` and `Transactional`), this block can only be reached with a `Received` delivery state, indicating partial transfer. I guess the latter case would be handled by the proton-j layer below and not be exposed here.

> It also feels strange that we finish the Span in this case but not in the other cases ...

In the other cases the span is finished by the `CommandContext#accept/reject/etc.` methods.